README.md
<p align="center"><img src="https://github.com/m-doescode/LuaStreams/raw/master/docs/logo.png"></p>
<h1 align="center">LuaStreams</h1>

(Icon designed by m-doescode using Lunacy created by [Icons8](https://icons8.com/))

LuaStreams brings the Java 8 Streams library to Lua or Luau (Roblox).

Streams are away to easily apply logic to a list without cluttering your code with many for-loops and if-statements.
e.g.
```lua
local t = {1, 6, 2, 3, 8}

table.sort(t)
local squared = {}
for i, num in ipairs(t) do
	squared[i] = math.pow(num, 2)
end
local randomNumber = math.random(1, #t)
print(randomNumber .. " of list is: " .. squared[i]) 
```
Can be written as:
```lua
local Streams = require(game.ServerScriptService.Streams)
local Stream = Streams.Stream

local t = {1, 6, 2, 3, 8}

local final = Stream.new(t)
		:sorted()
		:map(function(num) return math.pow(num, 2) end)
		:collect()

local randomNumber = math.random(1, #t)
print(randomNumber .. " of list is: " .. squared[i]) 
```

## Importing
### In plain-Luau projects (non-Roblox)
Use the same steps as for [in lua projects](#in-lua-projects) but use the Luau file at [Streams.lua](https://github.com/m-doescode/LuaStreams/blob/master/Streams.lua)

### In Roblox projects
1. Download the latest Luau file at [Streams.lua](https://github.com/m-doescode/LuaStreams/blob/master/Streams.lua)
2. Rename it to "Streams"
3. Put it in ServerScriptService
4. Require it via ```require(game.ServerScriptService.Streams)```

## Usage
### Basic Tutorial
If you have not used Java 8 Streams or you aren't completely sure how to use them, you can refer to here.

First, you must create a Stream via `Stream.new(t : table)` where t is a table that you want to use streams with.

Then you can chain function calls on the returned value like so:
```lua
Stream.new(t)
	:map(...)
	:sorted()
	:filter(...)
	...
```
Then finally, you may either use `forEach` to iterate over it or `collect` to convert it to a table.
e.g.
```lua
local t = {5, 2, 3, 1, 4}

Stream.new(t)
	:sorted(function(a, b) return a > b end)
	:forEach(function(num)
		print("Countdown: " .. num)
	end)
```
### Constructors
`Stream.new( t : { any } )` - Creates a new stream based on the list t.

`Stream.values( t : { any } )` - Gets the values from the dictionary t and creates a new Stream based on it.

`Stream.entrySet( t : { any } )` - Converts the dictionary t into a list of entries that go `{ key = <key>, value = <value> }`.

### Chain functions
`:filter( predicate : (any) -> boolean )` - Checks if the function returns true on each element. If it returns true, the element is kept, otherwise it is discarded.

`:map( mappingFunction : (any) -> any )` - Passes each element into the function and replaces it with the return value.

`:sorted()` - Sorts the table in ascending order according to `table.sort(t)`

`:sorted( comparator : (any, any) -> boolean )` - Sorts the table via `table.sort(t, comparator)`

`:limit( maxSize : number )` - Limits the size of the table to maxSize. If there are any more elements they are discarded.

`:skip( n : number )` - Skips first n elements in the table and discards them.

### Tail functions (final functions)
`:collect() : { any }` - Completes all the operations and converts the Stream into a table.

`:forEach( iterator : (any) -> nil )` - Iterates over the collected value of the Stream.

`:reduce<T>( identity : T, accumulator : (T, T) -> T )` - Passes each element as the second argument into the accumulator, and the last result (return value) as the first argument. If there was no last result, it will pass identity.

`:reduce<T>( accumulator : (T, T) -> T )` - Passes each element as the second argument into the accumulator, and the last result (return value) as the first argument. If there was no last result, the accumulator will not be called and result will be set to the first item in the table.

CHANGELOG
# Version 2.0

* Removed support for plain Lua as it was annoying to maintain. If you would like a plain Lua version, I'm happy to accept contributions.

# Version 1.0

(TBA)

test/main.lua
local Streams = require('../Streams')
local Stream = Streams.Stream

local printtable = require("printtable")

local oldt = {1, 2, 3, 4}
local t = Stream.new(oldt)
            :map(function(x) return x + 1 end)
            :collect()

printtable(oldt)
printtable(t)


return {}

Streams.lua
-- ## LUASTREAMS ##
-- Java 8 Streams in Luau
-- by majdTRM
-- Github: https://github.com/m-doescode/LuaStreams
																																																--[[
# Example #

local t = {
	4,
	10,
	50,
	500,
	230,
	160,
	300,
	20,
	15,
	60
}

local newTable = Stream.new(t)
					:sorted(function(a, b) return a < b end)
					:skip(5)
					:limit(4)
					:collect()

print(newTable)
-------------------
Output:
{
	60,
	160,
	230,
	300
}

# Usage #

local Streams = require(game.ServerScriptService.Streams) -- Replace with location of module
local Stream = Streams.Stream

Stream.new(t : { any }) -- Create a stream

Stream:collect() -- Finish the stream and return a new table

Stream:filter( predicate : (any) -> bool ) -- Filters the table based on the boolean return value. If it returns true, the value is kept.
If it returns false, the value is discarded.

Stream:map( mappingFunction : (any) -> any ) -- Converts (maps) each item in a table by passing it to the mappingFunction and replacing it with
the returned value.

type Comparator = (any, any) -> boolean

Stream:sorted( comparator : Comparator ) -- Sorts the table by passing each two values into the function. If the function returns true,
the first argument is put before, and the second argument is put after. If the function returns false, the second argument is put before,
and the first argument is put after. Equivalent to table.sort

Stream:limit( maxSize : number ) -- Limits the size of the table to maxSize. If there are more elements than maxSize, then those elements are
discarded.

Stream:skip( n : number ) -- Skips the first n elements in the table, discarding them.

Stream:forEach( action : (any) -> nil ) -- Iterates over the table, passing the value into the function every time. Equivalent to for-pairs loop.
																																						]]
-------------------

-- Helper Funcs

function doubleipairs(t)
	local function iter(t, i)
		i += 2
		local a = t[i]
		local b = t[i + 1]
		if a and b then
			return i, a, b
		end
	end
	return iter, t, -1
end

function comparingipairs(t)
	local function iter(t, i)
		i += 1
		local a = t[i]
		local b = t[i + 1]
		if a and b then
			return i, a, b
		end
	end
	return iter, t, 0
end

function shallowCopy(original : { any }) : { any }
	local copy = {}
	for key, value in pairs(original) do
		copy[key] = value
	end
	return copy
end

-- Classes

local Stream = {}
Stream.__index = Stream
type Stream = typeof(Stream)

function Stream.new(t : { any }) : Stream
	local self = {}

	self.table = t
	self.operations = setmetatable({}, {__index = table})

	return (setmetatable(self, Stream) :: any)
end

function Stream.values(t : { any }) : Stream
	local newt = setmetatable({}, {__index = table})

	for _, v in pairs(t) do
		newt:insert(v)
	end
	
	return Stream.new(newt)
end

function Stream.entrySet(t : { any }) : Stream
	local newt = setmetatable({}, {__index = table})

	for k, v in pairs(t) do
		newt:insert({ key = k, value = v })
	end
	
	return Stream.new(newt)
end

---

function Stream:collect(): { any }
	local t = self.table
	for _, operation in pairs(self.operations) do
		t = operation(t)
	end
	return t
end

function Stream:filter( predicate : (any) -> boolean ) : Stream
	self.operations:insert(function(t)
		local newt = setmetatable({}, {__index = table})
		for k, v in ipairs(t) do
			if predicate(v) then
				newt[k] = v
			end
		end
		return newt
	end)

	return self
end

function Stream:map( mappingFunction : (any) -> any ) : Stream
	self.operations:insert(function(t)
		local newt = setmetatable({}, {__index = table})
		for i, v in ipairs(t) do
			newt[i] = mappingFunction(v)
		end
		return newt
	end)

	return self
end

function Stream:sorted() : Stream
	self.operations:insert(function(t)
		local newt = setmetatable(shallowCopy(t), {__index = table})
		table.sort(newt)
		return newt	
	end)

	return self
end

type Comparator = (any, any) -> boolean

function Stream:sorted( comparator : Comparator ) : Stream
	self.operations:insert(function(t)
		local newt = setmetatable(shallowCopy(t), {__index = table})
		table.sort(newt, comparator)
		return newt	
	end)

	return self
end

function Stream:limit( maxSize : number ) : Stream
	self.operations:insert(function(t)
		local newt = setmetatable({}, {__index = table})
		for i = 1, math.min(#t, maxSize) do
			newt[i] = t[i]
		end
		return newt
	end)

	return self
end

function Stream:skip( n : number ) : Stream
	self.operations:insert(function(t)
		local newt = setmetatable(shallowCopy(t), {__index = table})
		for i = 1, math.max(1, n) do
			newt:remove(1)
		end
		return newt
	end)

	return self
end

--

function Stream:forEach( action : (any) -> nil )
	for i, v in ipairs(self:collect()) do
		action(v)
	end
end

function Stream:reduce<T>( identity : T, accumulator : (T) -> T )
	local result : T = identity
	for i, v in ipairs(self:collect()) do
		result = accumulator(result, v)
	end
	return result
end

function Stream:reduce<T>( accumulator : (T) -> T )
	local found : boolean = false
	local result : T
	for i, v in ipairs(self:collect()) do
		if not found then
			found = true
			result = v
		else
			result = accumulator(result, v)
		end
	end
	return result
end

local Streams = {}

Streams.Stream = Stream

return Streams

Streams - Lua.lua
-- ## LUASTREAMS ##
-- Java 8 Streams in Lua
-- by m-doescode
-- Github: https://github.com/m-doescode/LuaStreams
																																																--[[
# Example #

local t = {
	4,
	10,
	50,
	500,
	230,
	160,
	300,
	20,
	15,
	60
}

local newTable = Stream.new(t)
					:sorted(function(a, b) return a < b end)
					:skip(5)
					:limit(4)
					:collect()

print(newTable)
-------------------
Output:
{
	60,
	160,
	230,
	300
}

# Usage #

local Streams = require('libs/Streams.lua') -- Replace with location of module
local Stream = Streams.Stream

Stream.new(t : { any }) -- Create a stream

Stream:collect() -- Finish the stream and return a new table

Stream:filter( predicate : (any) -> bool ) -- Filters the table based on the boolean return value. If it returns true, the value is kept.
If it returns false, the value is discarded.

Stream:map( mappingFunction : (any) -> any ) -- Converts (maps) each item in a table by passing it to the mappingFunction and replacing it with
the returned value.

type Comparator = (any, any) -> boolean

Stream:sorted( comparator : Comparator ) -- Sorts the table by passing each two values into the function. If the function returns true,
the first argument is put before, and the second argument is put after. If the function returns false, the second argument is put before,
and the first argument is put after. Equivalent to table.sort

Stream:limit( maxSize : number ) -- Limits the size of the table to maxSize. If there are more elements than maxSize, then those elements are
discarded.

Stream:skip( n : number ) -- Skips the first n elements in the table, discarding them.

Stream:forEach( action : (any) -> nil ) -- Iterates over the table, passing the value into the function every time. Equivalent to for-pairs loop.
																																						]]
-------------------

-- Helper Funcs

function doubleipairs(t)
	local function iter(t, i)
		i += 2
		local a = t[i]
		local b = t[i + 1]
		if a and b then
			return i, a, b
		end
	end
	return iter, t, -1
end

function comparingipairs(t)
	local function iter(t, i)
		i += 1
		local a = t[i]
		local b = t[i + 1]
		if a and b then
			return i, a, b
		end
	end
	return iter, t, 0
end

function shallowCopy(original)
	local copy = {}
	for key, value in pairs(original) do
		copy[key] = value
	end
	return copy
end

-- Classes

local Stream = {}
Stream.__index = Stream

function Stream.new(t) : Stream
	local self = {}

	self.table = t
	self.operations = setmetatable({}, {__index = table})

	return setmetatable(self, Stream)
end

function Stream.values(t)
	local newt = setmetatable({}, {__index = table})

	for _, v in pairs(t) do
		newt:insert(v)
	end
	
	return Stream.new(newt)
end

function Stream.entrySet(t)
	local newt = setmetatable({}, {__index = table})

	for k, v in pairs(t) do
		newt:insert({ key = k, value = v })
	end
	
	return Stream.new(newt)
end

---

function Stream:collect()
	local t = self.table
	for _, operation in pairs(self.operations) do
		t = operation(t)
	end
	return t
end

function Stream:filter( predicate )
	self.operations:insert(function(t)
		local newt = setmetatable({}, {__index = table})
		for k, v in ipairs(t) do
			if predicate(v) then
				newt[k] = v
			end
		end
		return newt
	end)

	return self
end

function Stream:map( mappingFunction )
	self.operations:insert(function(t)
		local newt = setmetatable({}, {__index = table})
		for i, v in ipairs(t) do
			newt[i] = mappingFunction(v)
		end
		return newt
	end)

	return self
end

function Stream:sorted()
	self.operations:insert(function(t)
		local newt = setmetatable(shallowCopy(t), {__index = table})
		table.sort(newt)
		return newt	
	end)

	return self
end

function Stream:sorted( comparator )
	self.operations:insert(function(t)
		local newt = setmetatable(shallowCopy(t), {__index = table})
		table.sort(newt, comparator)
		return newt	
	end)

	return self
end

function Stream:limit( maxSize )
	self.operations:insert(function(t)
		local newt = setmetatable({}, {__index = table})
		for i = 1, math.min(#t, maxSize) do
			newt[i] = t[i]
		end
		return newt
	end)

	return self
end

function Stream:skip( n )
	self.operations:insert(function(t)
		local newt = setmetatable(shallowCopy(t), {__index = table})
		for i = 1, math.max(1, n) do
			newt:remove(1)
		end
		return newt
	end)

	return self
end

--

function Stream:forEach( action )
	for i, v in ipairs(self:collect()) do
		action(v)
	end
end

function Stream:reduce( identity, accumulator)
	local result = identity
	for i, v in ipairs(self:collect()) do
		result = accumulator(result, v)
	end
	return result
end

function Stream:reduce( accumulator )
	local found = false
	local result
	for i, v in ipairs(self:collect()) do
		if not found then
			found = true
			result = v
		else
			result = accumulator(result, v)
		end
	end
	return result
end

local Streams = {}

Streams.Stream = Stream

return Streams

Streams - Luau-Roblox.lua
-- ## LUASTREAMS ##
-- Java 8 Streams in Luau
-- by m-doescode
-- Github: https://github.com/m-doescode/LuaStreams
																																																--[[
# Example #

local t = {
	4,
	10,
	50,
	500,
	230,
	160,
	300,
	20,
	15,
	60
}

local newTable = Stream.new(t)
					:sorted(function(a, b) return a < b end)
					:skip(5)
					:limit(4)
					:collect()

print(newTable)
-------------------
Output:
{
	60,
	160,
	230,
	300
}

# Usage #

local Streams = require(game.ServerScriptService.Streams) -- Replace with location of module
local Stream = Streams.Stream

Stream.new(t : { any }) -- Create a stream

Stream:collect() -- Finish the stream and return a new table

Stream:filter( predicate : (any) -> bool ) -- Filters the table based on the boolean return value. If it returns true, the value is kept.
If it returns false, the value is discarded.

Stream:map( mappingFunction : (any) -> any ) -- Converts (maps) each item in a table by passing it to the mappingFunction and replacing it with
the returned value.

type Comparator = (any, any) -> boolean

Stream:sorted( comparator : Comparator ) -- Sorts the table by passing each two values into the function. If the function returns true,
the first argument is put before, and the second argument is put after. If the function returns false, the second argument is put before,
and the first argument is put after. Equivalent to table.sort

Stream:limit( maxSize : number ) -- Limits the size of the table to maxSize. If there are more elements than maxSize, then those elements are
discarded.

Stream:skip( n : number ) -- Skips the first n elements in the table, discarding them.

Stream:forEach( action : (any) -> nil ) -- Iterates over the table, passing the value into the function every time. Equivalent to for-pairs loop.
																																						]]
-------------------

-- Helper Funcs

function doubleipairs(t)
	local function iter(t, i)
		i += 2
		local a = t[i]
		local b = t[i + 1]
		if a and b then
			return i, a, b
		end
	end
	return iter, t, -1
end

function comparingipairs(t)
	local function iter(t, i)
		i += 1
		local a = t[i]
		local b = t[i + 1]
		if a and b then
			return i, a, b
		end
	end
	return iter, t, 0
end

function shallowCopy(original : { any }) : { any }
	local copy = {}
	for key, value in pairs(original) do
		copy[key] = value
	end
	return copy
end

-- Classes

local Stream = {}
Stream.__index = Stream
type Stream = typeof(Stream)

function Stream.new(t : { any }) : Stream
	local self = {}

	self.table = t
	self.operations = setmetatable({}, {__index = table})

	return (setmetatable(self, Stream) :: any)
end

function Stream.values(t : { any }) : Stream
	local newt = setmetatable({}, {__index = table})

	for _, v in pairs(t) do
		newt:insert(v)
	end
	
	return Stream.new(newt)
end

function Stream.entrySet(t : { any }) : Stream
	local newt = setmetatable({}, {__index = table})

	for k, v in pairs(t) do
		newt:insert({ key = k, value = v })
	end
	
	return Stream.new(newt)
end

---

function Stream:collect(): { any }
	local t = self.table
	for _, operation in pairs(self.operations) do
		t = operation(t)
	end
	return t
end

function Stream:filter( predicate : (any) -> boolean ) : Stream
	self.operations:insert(function(t)
		local newt = setmetatable({}, {__index = table})
		for k, v in ipairs(t) do
			if predicate(v) then
				newt[k] = v
			end
		end
		return newt
	end)

	return self
end

function Stream:map( mappingFunction : (any) -> any ) : Stream
	self.operations:insert(function(t)
		local newt = setmetatable({}, {__index = table})
		for i, v in ipairs(t) do
			newt[i] = mappingFunction(v)
		end
		return newt
	end)

	return self
end

function Stream:sorted() : Stream
	self.operations:insert(function(t)
		local newt = setmetatable(shallowCopy(t), {__index = table})
		table.sort(newt)
		return newt	
	end)

	return self
end

type Comparator = (any, any) -> boolean

function Stream:sorted( comparator : Comparator ) : Stream
	self.operations:insert(function(t)
		local newt = setmetatable(shallowCopy(t), {__index = table})
		table.sort(newt, comparator)
		return newt	
	end)

	return self
end

function Stream:limit( maxSize : number ) : Stream
	self.operations:insert(function(t)
		local newt = setmetatable({}, {__index = table})
		for i = 1, math.min(#t, maxSize) do
			newt[i] = t[i]
		end
		return newt
	end)

	return self
end

function Stream:skip( n : number ) : Stream
	self.operations:insert(function(t)
		local newt = setmetatable(shallowCopy(t), {__index = table})
		for i = 1, math.max(1, n) do
			newt:remove(1)
		end
		return newt
	end)

	return self
end

--

function Stream:forEach( action : (any) -> nil )
	for i, v in ipairs(self:collect()) do
		action(v)
	end
end

function Stream:reduce<T>( identity : T, accumulator : (T) -> T )
	local result : T = identity
	for i, v in ipairs(self:collect()) do
		result = accumulator(result, v)
	end
	return result
end

function Stream:reduce<T>( accumulator : (T) -> T )
	local found : boolean = false
	local result : T
	for i, v in ipairs(self:collect()) do
		if not found then
			found = true
			result = v
		else
			result = accumulator(result, v)
		end
	end
	return result
end

local Streams = {}

Streams.Stream = Stream

return Streams